[Round1] Step3 - Moss

0-jerry opened the pull request:
@lxodud 

**빼빼로의 기본형이 있으면 좋겠다는 생각에 
형태를 입력받는 문자열 매개변수 styleOfPepero와 topping은 기본값을 지정해주었습니다.**
>>이에 대해 궁금했던 점
기본값이 지정되어 있는 매개변수는 뒤쪽으로 위치시키는 것이 좋다 하셔서 정보에 출력되는 순서와 함수에서 입력받는 순서가 불일치하게 됐습니다.
만약 정보에 출력되는 순서가 맥락상 자연스럽다면 
함수에서 출력되는 매개변수의 순서를 위해서 
기본값을 포기하는 방식과 기본값이 지정되었더라도 출력되는 순서에 일치시켜 배치하는 것 중에 
어떤 것이 더 일반적일까요?

0-jerry committed:
Step.3 빼빼로 그리기 코드 추가

0-jerry committed:
순서도 추가

lxodud reviewed (commented):
@lichen0423
고생하셨습니다 👍👍👍

>이에 대해 궁금했던 점
기본값이 지정되어 있는 매개변수는 뒤쪽으로 위치시키는 것이 좋다 하셔서 정보에 출력되는 순서와 함수에서 입력받는 순서가 불일치하게 됐습니다.
만약 정보에 출력되는 순서가 맥락상 자연스럽다면
함수에서 출력되는 매개변수의 순서를 위해서
기본값을 포기하는 방식과 기본값이 지정되었더라도 출력되는 순서에 일치시켜 배치하는 것 중에
어떤 것이 더 일반적일까요?

개인적으로 내부에서 출력되는 순서와 파라미터의 순서를 동일하게 일치시키는게 크게 중요하다고 생각하지 않습니다. 물론 사람에 따라 다르게 생각할 수 있는 지점입니다.
[Swift - API Design Guidelines](https://www.swift.org/documentation/api-design-guidelines/)를 읽어보시고 어떠한 이유에서 네이밍을 했는지에 대한 근거를 만들면 문제없다고 생각합니다.

0-jerry committed:
코드수정

0-jerry committed:
modify-code

1 .`let formOfPepero: String = topping + styleOfPepero + topping.reversed()` 에서
topping.reversed() 가 Array<Character>타입으로 인식되어 String 타입으로 변환해주었습니다. 

2. todo 양식에 맞춰 메서드를 구현한 뒤, 메서드들 호출해주는 함수를 정의하는 방식으로 수정했습니다.

lxodud reviewed (approved):
>위 부분에서 궁금한 점이있습니다.
let formOfPepero: String = topping + styleOfPepero
formOfPepero += topping.reversed()
로 나눠서 실행시키면 위 오류가 사라지는데 혹시 어떤 이유로 달라지는 걸까요?
혹시 let formOfPepero: String = topping + styleOfPepero + topping.reversed() 의 경우에는
topping.reversed()가 Character 의 집합이기에 타입 불일치로 연산이 허용되지 않지만,
formOfPepero += topping.reversed() 에선 String 타입의 요소인 Character들을 연산하는 형태이기에 허용되는 걸까요?

이런 의문을 해소하기 위해서 해당 연산자가 어떻게 구현되어있는지 보는게 좋습니다.
[+=](https://developer.apple.com/documentation/swift/string/+=(_:_:)-676gx)
위 공식문서의 메서드 시그니처를 보면 아래와 같습니다.
```swift
static func += <Other>(lhs: inout Self, rhs: Other) where Other : Sequence, Self.Element== Other.Element
```
제네릭 문법이 사용되고 있어서 조금 어려울 수 있지만 질문주신 코드에 대입해서 얘기해보겠습니다.
먼저 lhs 가 String이겠죠 그리고 rhs가 Other인데 where뒤를 보면 이건 Sequence라는 프로토콜을 채택하는 타입이라는 제약이 있습니다. 아무튼 저 Other가 ReversedCollection<Self>이라는 타입이겠죠 그리고 제일 마지막에 `Self.Element== Other.Element`가 의미하는 바가 제일 중요한데 lsh와 rhs의 Element가 동일하다는 제약입니다. 위에서 질문주신 코드를 보시면 lhs(String), rhs(ReversedCollection)의 Element가 동일하겠죠 따라서 연산이 되는 것입니다!
어려울 수 있는 개념이라 지금 완벽하게 이해안하고 넘어가도 괜찮은 주제라고 생각합니다.

고생하셨습니다 👍👍👍
